### PR TITLE
[LLVM][IR] When evaluating GEP offsets don't assume ConstantInt is a scalar.

### PR DESCRIPTION
**Pull request by paulwalker-arm**

### CODE REVIEW

File: llvm/lib/IR/Operator.cpp [hunk 1]
```diff
@@ -163,28 +165,30 @@ bool GEPOperator::accumulateConstantOffset(
...
-      // Handle a struct index, which adds its field offset to the pointer.
-      if (STy) {
-        unsigned ElementIdx = ConstOffset->getZExtValue();
-        const StructLayout *SL = DL.getStructLayout(STy);
-        // Element offset is in bytes.
-        if (!AccumulateOffset(
-                APInt(Offset.getBitWidth(), SL->getElementOffset(ElementIdx)),
-                1))
+    if (V->getType()->isIntegerTy()) {
+      if (auto ConstOffset = dyn_cast<ConstantInt>(V)) {
```
Comment (nikic) [inline llvm/lib/IR/Operator.cpp:169]: In the interest of keeping the indentation down, I'd write this as:
```
auto *ConstOffset = dyn_cast<ConstantInt>(V);
if (ConstOffset && ConstOffset->getType()->isIntegerTy()) {
}
```

File: llvm/lib/IR/Operator.cpp [hunk 2]
```diff
@@ -163,28 +165,30 @@ bool GEPOperator::accumulateConstantOffset(
...
+          const StructLayout *SL = DL.getStructLayout(STy);
+          // Element offset is in bytes.
+          if (!AccumulateOffset(
+                  APInt(Offset.getBitWidth(), SL->getElementOffset(ElementIdx)),
+                  1))
+            return false;
+          continue;
+        }
+        if (!AccumulateOffset(ConstOffset->getValue(),
+                              GTI.getSequentialElementStride(DL)))
```
Comment (nikic) [inline llvm/lib/IR/Operator.cpp:188]: Also update the very similar code in collectOffset() below?

Comment (paulwalker-arm) [reply llvm/lib/IR/Operator.cpp]: Done.  I've been trying to rely purely on existing testing but it seems this was not covered so I've added a new one.

**Review by nikic — APPROVED** [review]
LGTM. We probably *should* support vectors here in the future, but that's going to have fallout, so this is fine for now...

**Pull request merged**